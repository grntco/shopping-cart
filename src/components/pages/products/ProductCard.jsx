import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import starIcon from '../../../assets/icons/star.svg'
import Icon from '../../reusables/Icon'
import Button from '../../reusables/Button'
import { useState } from 'react'

const ProductCard = ({ product, handleAddToCart }) => {
    const [quantity, setQuantity] = useState(0)

    return (
        <StyledCard>
            <CardImageLink>
                <img
                    src={product.image}
                    alt={product.title}
                    title={product.title}
                />
            </CardImageLink>
            <CardInfoContainer>
                <CardInfo>
                    <Link>
                        <h4>{product.title}</h4>
                    </Link>
                    <CardInfoRating>
                        <Icon src={starIcon} alt='star icon' />
                        <p>{product.rating.rate}</p>
                    </CardInfoRating>
                </CardInfo>
                <CardInfo>
                    <div>
                        <Button onClick={handleDecrementQuantity}>-</Button>
                        <Input value={quantity} />
                        <Button onClick={handleIncrementQuantity}>+</Button>
                    </div>
                    <p>${product.price}</p>
                </CardInfo>
                <Button
                    onClick={() => {
                        handleAddToCart(product)
                    }}
                >
                    Add to Cart
                </Button>
            </CardInfoContainer>
        </StyledCard>
    )

    function handleIncrementQuantity() {
        setQuantity(quantity + 1)
    }

    function handleDecrementQuantity() {
        if (quantity > 0) setQuantity(quantity - 1)
    }
}

const StyledCard = styled.div`
    border: 1px solid black;
    display: flex;
    flex-direction: column;
    padding: 24px;
`

const CardImageLink = styled(Link)`
    align-self: center;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-grow: 1;
    width: 100%;
    max-width: 336px;

    img {
        width: 100%;
    }
`

const CardInfoContainer = styled.div`
    display: flex;
    flex-direction: column;
`

const CardInfo = styled.div`
    display: flex;
    justify-content: space-between;
`

const Input = styled.input`
    &::-webkit-inner-spin-button,
    &::-webkit-outer-spin-button {
        -webkit-appearance: none;
        margin: 0;
    }
`

const CardInfoRating = styled.div`
    display: flex;
    flex-wrap: nowrap;
`

ProductCard.propTypes = {
    product: PropTypes.object.isRequired,
    handleAddToCart: PropTypes.func.isRequired,
}

export default ProductCard
