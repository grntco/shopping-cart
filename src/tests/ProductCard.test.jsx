import { describe, it, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
// import Header from '../components/Header'
// import Router from '../components/Router'
// import userEvent from '@testing-library/user-event'
// import App from '../components/App'
import ProductCard from '../components/pages/products/ProductCard'
import { BrowserRouter } from 'react-router-dom'

const mockProduct = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
        'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
        rate: 3.9,
        count: 120,
    },
}

describe('Product Card test', () => {
    it('Product card renders correctly', () => {
        render(
            <ProductCard product={mockProduct} handleAddToCart={() => {}} />,
            { wrapper: BrowserRouter },
        )
        expect(screen.getByTestId('product-card')).toBeInTheDocument()
        expect(screen.getByTestId('product-card')).toMatchSnapshot()
    })
})
