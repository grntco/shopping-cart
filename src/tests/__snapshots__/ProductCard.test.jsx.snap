// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Card test > Product card renders correctly 1`] = `
<div
  class="sc-eDLKkx ceMjzl"
  data-testid="product-card"
>
  <a
    class="sc-jTQCzO imamgb"
    data-testid="product-card__image-link"
    href="/products/1"
  >
    <img
      alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
      src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      title="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
    />
  </a>
  <div
    class="sc-gLLuof kTfTWK"
  >
    <div
      class="sc-iBdnpw jphUIh"
    >
      <a
        data-testid="product-card__title-link"
        href="/products/1"
      >
        <h4>
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </h4>
      </a>
      <div
        class="sc-fsYfdN cwoSHM"
      >
        <img
          alt="star icon"
          class="sc-beySPh htjDQV"
          src="/src/assets/icons/star.png"
        />
        <p>
          3.9
        </p>
      </div>
    </div>
    <p>
      $109.95
    </p>
    <div
      class="sc-dmyCSP gUwTfa"
    >
      <div
        class="sc-hLQSwg tApZJ"
      >
        <button
          class="sc-guDLey hYdscr"
        >
          â€“
        </button>
        <input
          type="number"
          value="1"
        />
        <button
          class="sc-guDLey hYdscr"
        >
          +
        </button>
      </div>
      <button
        class="sc-guDLey hYdscr"
      >
        Add to Cart
      </button>
    </div>
  </div>
</div>
`;
